### ISSUE TITLE
Remove information used for labelling (overfitting prevention)

### ISSUE THREAD

**Issue by chrisji** [snorkel-team/snorkel#1423]
Hi - could not see any related issues, but forgive me if this already has a thread

When labelling instances, is there a way to transform the data to remove information used during labelling? 

Trivial example:
```
"5 stars, I really like this product"
"This product is pretty bad 2/5"
"4 stars, product does the job but is expensive"
...
```

For a labelling function that extracts ratings using some regexes e.g. `[0-5] star(s?)` and `[0-5]/5`, it would be important that the dataset is augmented to prevent the training data from containing the class label (overfitting would occur):

```
", I really like this product"
"this product is pretty bad"
", product does the job but is expensive"
...
```

Obviously, this is a trivial post-labelling step when there is one labelling function, but when the annotation may come from multiple labelling functions one may like to specify how to augment the data based on the labeller used.

Is there a more elegant solution than applying a function that destroys all possible information the labellers _may_ have used? Perhaps an "if labelled using X, then transform using Y"?

**Comment by vincentschen** [comment]
Hi @chrisji, interesting problem! Absolutely — you could use [transformation functions](https://snorkel.readthedocs.io/en/v0.9.0/packages/_autosummary/augmentation/snorkel.augmentation.transformation_function.html#snorkel.augmentation.transformation_function) to perhaps remove the substrings where ratings are extracted. This tutorial might be helpful: https://www.snorkel.org/use-cases/02-spam-data-augmentation-tutorial 

It might also be worth checking our [forum](https://spectrum.chat/snorkel) to poll if anyone is trying something similar.

**Comment by ajratner** [comment]
@chrisji this is a very interesting question!  And I agree that transformation functions of some kind or other might be a great solution here!

Worth explicitly noting though that more broadly, the general baseline solution to the above problem (more broadly than Snorkel- in any ML setting) where there are some *highly predictive* features that you don't want to overfit to, is to use regularization.  So I would make sure to see how far just basic e.g. L2 regularization gets you (and of course, note that transformation functions are themselves a type of regularizer)

**Comment by mostlymetalman** [comment]
@chrisji just to build on the above, and assuming your task is language-based, rather than removing the information with a TF you might find random replacement-based transformations, like those in the spam tutorial, work better. The advantage of doing this is that you can avoid making the training data fundamentally different than real data (if you delete, there will be phrases it sees when you use the model it never could have seen in the training data and inputs with entirely different syntactic structures). Sure, the inputs then become semantically weird ("1 star, I really like this product"...though I bet we've all seen reviews like this too) but you accomplish your goal of discouraging overfitting while making sure all the right types of words and symbols are in the right places. 

I have not done this with snorkel but it seems to be a straightforward application of TFs. I've been thinking some about it though. I realize you are asking for an automated solution but maybe just making sure that when you write an LF you write a corresponding TF that goes and noises up what the LF would use? Personally, I don't think I'd want this to be automated since I'd want to control the semantics of how the data gets transformed, and just deleting everything that gets used may corrupt the input too much.
